src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -111,13 +118,49 @@ private static IComponentActivator GetComponentActivatorOrDefault(IServiceProvid
         /// </summary>
         protected internal ElementReferenceContext? ElementReferenceContext { get; protected set; }
 
+        internal HotReloadContext HotReloadContext { get; } = new();
+
+        private async void RenderRootComponentsOnHotReload()
+        {
+            await Dispatcher.InvokeAsync(async () =>
+            {
+                Debug.Assert(_rootComponents is not null);
+                HotReloadContext.IsHotReloading = true;
+
+                try
+                {
+                    _pendingTasks = new List<Task>();
+                    foreach (var (componentState, initialParameters) in _rootComponents)
+                    {
+                        AddToPendingTasks(componentState.Component.SetParametersAsync(initialParameters));
+                    }
+
+                    await ProcessAsynchronousWork();
+                    Debug.Assert(_pendingTasks.Count == 0);
+                }
```

Comment:
This also brings up the point that, in some applications, hot reload refreshes might not be as state-preserving as you'd want. Consider if a root (or deep) component always does some async data loading while showing `Loading...` instead of its child content, then instantiates its descendant subtree. In that case, each code delta is going to retrigger this flow, and you won't get the instant update in the UI you're actually working on (you'll see "Loading..."), and any transient state in descendants will be discarded.

**I'm not disputing that you're doing the right thing by re-rendering the whole tree** (I'm disputing the async thing, but that's separate). I'm just realising that while our demo cases hot-reloaded really well, there will be some other cases that don't. It will be up to app developers to make their component hierarchies refresh in a nondestructive way to get the best experience. For example they should try to do data loading in `OnInitializedAsync` rather than `OnParametersSetAsync` (except if they manually compare old and new parameters) so that each hot reload doesn't rebuild the descendant tree from scratch. We should think about how to set this expectation in how we document and communicate the feature (cc @danroth27).